core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/ArrayBindingBasedQueryEvaluationContext.java

```diff
@@ -23,25 +25,56 @@
 import org.eclipse.rdf4j.query.MutableBindingSet;
 import org.eclipse.rdf4j.query.QueryEvaluationException;
 import org.eclipse.rdf4j.query.algebra.ExtensionElem;
+import org.eclipse.rdf4j.query.algebra.Group;
 import org.eclipse.rdf4j.query.algebra.MultiProjection;
+import org.eclipse.rdf4j.query.algebra.Projection;
 import org.eclipse.rdf4j.query.algebra.ProjectionElem;
 import org.eclipse.rdf4j.query.algebra.QueryRoot;
+import org.eclipse.rdf4j.query.algebra.StatementPattern;
+import org.eclipse.rdf4j.query.algebra.UnaryTupleOperator;
 import org.eclipse.rdf4j.query.algebra.Var;
 import org.eclipse.rdf4j.query.algebra.ZeroLengthPath;
 import org.eclipse.rdf4j.query.algebra.evaluation.ArrayBindingSet;
 import org.eclipse.rdf4j.query.algebra.evaluation.iterator.ZeroLengthPathIteration;
-import org.eclipse.rdf4j.query.algebra.helpers.AbstractQueryModelVisitor;
+import org.eclipse.rdf4j.query.algebra.helpers.AbstractSimpleQueryModelVisitor;
+import org.eclipse.rdf4j.query.impl.EmptyBindingSet;
 
 public final class ArrayBindingBasedQueryEvaluationContext implements QueryEvaluationContext {
 	private final QueryEvaluationContext context;
 	private final String[] allVariables;
 	private final LinkedHashSet<String> allVariablesSet;
+	private final ArrayBindingSet defaultArrayBindingSet;
+	private final Predicate<BindingSet>[] hasBinding;
+	private final Function<BindingSet, Binding>[] getBinding;
+	private final Function<BindingSet, Value>[] getValue;
+	private final BiConsumer<Value, MutableBindingSet>[] setBinding;
+	private final BiConsumer<Value, MutableBindingSet>[] addBinding;
```

Comment:
These are caches for the generated lambdas. This is faster because it limits the number of lambdas that the compiler has to compile through C1 to C2.